File: backend/schemas/timer.py
"""专注计时（openapi.yaml `/timer-sessions` 系列）

TimerSession / TimerSegment / TimerSessionStart / TimerSegmentStart /
TimerRestore / TimerCurrent / TimerFinish / TimerDiscarded

字段名与枚举取值以 docs/openapi.yaml 为准（TimerSession / TimerSegment / TimerRestore
已在 M0 冻结，本模块补齐请求体与"当前会话"包装）。
"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from .enums import Completion, Subject, TimerMode, TimerStatus
from .learning_record import RecordSelfReport


class TimerSegment(BaseModel):
    """计时会话内的一段（#14 分段计时）。一个会话可挂多段，每段标注所属任务。"""

    model_config = ConfigDict(populate_by_name=True)

    segment_id: str = Field(..., alias="segmentId")
    task_id: str | None = Field(None, alias="taskId", description="本段属于哪个任务；切换任务即新开一段")
    started_at: datetime = Field(..., alias="startedAt")
    ended_at: datetime | None = Field(None, alias="endedAt")
    seconds: int | None = Field(None, description="本段有效秒数（口径同 effectiveSeconds）")


class TimerSession(BaseModel):
    """计时会话（服务端持久化的"进行中"记录 = 专注态的唯一真相源）。"""

    model_config = ConfigDict(populate_by_name=True)

    session_id: str = Field(..., alias="sessionId")
    mode: TimerMode
    started_at: datetime = Field(..., alias="startedAt", description="两种模式都有；恢复计算的基准")
    target_minutes: int | None = Field(None, alias="targetMinutes", description="仅 countdown 有；countup 为 null")
    plan_id: str | None = Field(None, alias="planId")
    task_id: str | None = Field(None, alias="taskId")
    subject: Subject
    status: TimerStatus
    ended_at: datetime | None = Field(None, alias="endedAt")
    effective_seconds: int | None = Field(
        None, alias="effectiveSeconds", description="**有效时长**（秒），≠ 墙上时长（D31）"
    )
    last_heartbeat_at: datetime | None = Field(None, alias="lastHeartbeatAt")
    segments: list[TimerSegment] = Field(default_factory=list, description="分段（#14）")
    created_at: datetime = Field(..., alias="createdAt")


class TimerSessionStart(BaseModel):
    """开始一次计时。

    `targetMinutes` 仅 countdown 必填；countup 传了会被忽略（不报错，避免前端切换模式时多余的 400）。
    `subject` 可不传——服务端会从 taskId 对应的计划任务推导，推不出来才兜底 `other`。
    """

    model_config = ConfigDict(
        populate_by_name=True,
        json_schema_extra={
            "example": {"mode": "countdown", "targetMinutes": 30, "planId": "p_1a2b", "taskId": "t_3c4d"}
        },
    )

    mode: TimerMode
    target_minutes: int | None = Field(None, alias="targetMinutes", ge=1, le=600)
    plan_id: str | None = Field(None, alias="planId")
    task_id: str | None = Field(None, alias="taskId")
    subject: Subject | None = Field(None, description="不传则从 taskId 推导，推不出兜底 other")


class TimerSegmentStart(BaseModel):
    """切换任务 → 开新的一段（#14：一会话内分段，不允许并行计时）。"""

    model_config = ConfigDict(populate_by_name=True)

    task_id: str | None = Field(None, alias="taskId", description="本段要做什么任务；不传表示自由学习")


class TimerRestore(BaseModel):
    """恢复视图（D30 按 mode 分支计算）。

    `needsVerdict=true` 表示识别到**异常会话**，前端必须弹**恢复裁决卡**（D31），**不得自动记账**。
    """

    model_config = ConfigDict(populate_by_name=True)

    session_id: str = Field(..., alias="sessionId")
    mode: TimerMode
    remaining_seconds: int | None = Field(
        None, alias="remainingSeconds", description="仅 countdown：剩余 = target − 已过；≤0 视为已到点（软提醒，不自动结束）"
    )
    elapsed_seconds: int | None = Field(None, alias="elapsedSeconds", description="仅 countup：累计 = now − startedAt")
    needs_verdict: bool = Field(..., alias="needsVerdict", description="是否需用户裁决（断连 / 超上限）")
    suggested_minutes: int | None = Field(
        None, alias="suggestedMinutes", description="裁决卡「保留按 X 记」的预填值（倒计时封顶 target；正计时按截断点）"
    )


class TimerCurrent(BaseModel):
    """`GET /timer-sessions/current`：当前是否有进行中的会话。

    用 `active` 包装而不是"没有就 404"：前端每次进计时页都会问一次，
    "没有会话"是**正常流程**而不是错误，不该走错误分支。
    """

    model_config = ConfigDict(populate_by_name=True)

    active: bool
    session: TimerSession | None = None
    restore: TimerRestore | None = None


class TimerFinish(BaseModel):
    """计时收尾（D20 三层形态：结束瞬间 0 步 → 轻收尾卡 → 正确率延后补）。

    必填只有 `completion`（唯一"半强制"问句）；`selfReport` **可整段省略**、
    其中专注/疲劳/情绪/难度**任一可缺**（软字段转译不出就留空，D34 不造数）。
    `durationMinutes` 显式传入时**覆盖服务端按 mode 算出的有效时长**——
    这正是裁决卡第三态「手动改时长」的落点。
    """

    model_config = ConfigDict(
        populate_by_name=True,
        json_schema_extra={
            "example": {
                "completion": "completed",
                "selfReport": {"focus": 4, "fatigue": 2, "emotion": "positive", "difficultyFeel": "moderate"},
                "note": "函数图像终于看懂了",
            }
        },
    )

    completion: Completion
    self_report: RecordSelfReport | None = Field(
        None, alias="selfReport", description="自评；可整段省略，软字段也可缺"
    )
    duration_minutes: int | None = Field(
        None, alias="durationMinutes", ge=1, le=600, description="不传则按 mode 算有效时长；传入即以它为准"
    )
    note: str | None = Field(None, max_length=100, description="≤100 字备注")
    skip_recommendation: bool | None = Field(None, alias="skipRecommendation")


class TimerDiscarded(BaseModel):
    """裁决卡「丢弃」的结果。**不产生学习记录**（D31 验收：僵尸会话不产生记录）。"""

    model_config = ConfigDict(populate_by_name=True)

    discarded: bool
    session_id: str = Field(..., alias="sessionId")

File: frontend/src/services/timer.ts
/**
 * 专注计时（服务端会话）—— D30 / D31 / D32 / #14。
 *
 * 接口映射（docs/openapi.yaml v1.7.1）：
 * - POST   /timer-sessions                       开始计时（已有进行中会话时 409）
 * - GET    /timer-sessions/current               当前会话 + 恢复视图（active=false 是正常流程）
 * - GET    /timer-sessions/{sessionId}           会话详情
 * - POST   /timer-sessions/{sessionId}/heartbeat 上报心跳（僵尸判定的唯一依据）
 * - POST   /timer-sessions/{sessionId}/segments  切换任务（开新的一段，#14）
 * - POST   /timer-sessions/{sessionId}/finish    结束并收尾（原子落学习记录）
 * - DELETE /timer-sessions/{sessionId}           丢弃（裁决卡「丢弃」，不产生记录）
 *
 * **为什么计时状态在服务端**：原来 `/study-timer` 的上下文靠 react-router 的
 * `navigate(state)` 传参，刷新或直链进来就丢任务、丢时长、丢 planId。
 * 目标态 §3.6 定的口径是「服务端持久化进行中的计时会话，前端仅是视图」——
 * 所以本模块是"专注态的唯一真相源"，前端回来只需 `getCurrentTimerSession()` 按 mode 恢复。
 *
 * ⚠️ 类型定义放在本文件而**不是** `types/api.ts`：后者按板块契约归 X1 管
 * （X0 定字段 / X1 管导出），C 板块不动手。等 X1 收口时把这里的定义合并过去即可。
 */

import { apiDelete, apiGet, apiPost } from './http';
import type { Subject } from '@/types/api';

/** openapi.yaml TimerMode */
export type TimerMode = 'countdown' | 'countup';
/** openapi.yaml TimerStatus */
export type TimerStatus = 'running' | 'finished' | 'abandoned';
/** openapi.yaml Completion */
export type Completion = 'completed' | 'partial' | 'abandoned';
/** openapi.yaml Emotion */
export type Emotion = 'positive' | 'neutral' | 'negative';
/** openapi.yaml DifficultyFeel */
export type DifficultyFeel = 'easy' | 'moderate' | 'hard';

/**
 * 自评（openapi.yaml RecordSelfReport v1.7.1）。
 *
 * ⚠️ 四字段**全部可空**：三层收尾里唯一"半强制"的只有完成度；专注/疲劳/难度是
 * 模型从"一句感受"转译的软字段，情绪快捷词也只是"可选兜底"。转译不出就留空——
 * 服务端会跳过该项并按可用部分归一化，**不造数**（D34）。
 */
export interface SelfReportInput {
  focus?: number | null;
  fatigue?: number | null;
  emotion?: Emotion | null;
  difficultyFeel?: DifficultyFeel | null;
}

export interface TimerSegment {
  segmentId: string;
  taskId?: string | null;
  startedAt: string;
  endedAt?: string | null;
  seconds?: number | null;
}

export interface TimerSession {
  sessionId: string;
  mode: TimerMode;
  startedAt: string;
  targetMinutes?: number | null;
  planId?: string | null;
  taskId?: string | null;
  subject: Subject;
  status: TimerStatus;
  endedAt?: string | null;
  /** 有效时长（秒）—— ≠ 墙上时长；倒计时封顶 target */
  effectiveSeconds?: number | null;
  lastHeartbeatAt?: string | null;
  segments: TimerSegment[];
  createdAt: string;
}

/**
 * 恢复视图（D30 按 mode 分支计算）。
 * `needsVerdict=true` 表示识别到**异常会话**，必须弹**恢复裁决卡**（D31），**不得自动记账**。
 */
export interface TimerRestore {
  sessionId: string;
  mode: TimerMode;
  /** 仅 countdown：剩余 = target − 已过；≤0 视为已到点（软提醒，不自动结束） */
  remainingSeconds?: number | null;
  /** 仅 countup：累计 = now − startedAt */
  elapsedSeconds?: number | null;
  needsVerdict: boolean;
  /** 裁决卡「保留按 X 记」的预填值 */
  suggestedMinutes?: number | null;
}

export interface TimerCurrent {
  active: boolean;
  session?: TimerSession | null;
  restore?: TimerRestore | null;
}

export interface TimerSessionStart {
  mode: TimerMode;
  /** 仅 countdown 需要；countup 传了会被忽略（不报错） */
  targetMinutes?: number;
  planId?: string;
  taskId?: string;
  subject?: Subject;
}

export interface TimerFinishInput {
  completion: Completion;
  selfReport?: SelfReportInput;
  /**
   * 显式传入时**覆盖**服务端按 mode 算出的有效时长。
   * 前端只在"有过暂停"时传（服务端不知道前端暂停了多久）；
   * 无暂停时留空，让服务端按 mode 算——倒计时的封顶逻辑在服务端更可靠。
   */
  durationMinutes?: number;
  note?: string;
  skipRecommendation?: boolean;
}

/** 提交记录后返回的建议句柄（轮询 GET /recommendations/{id} 用） */
export interface RecommendationHandle {
  recommendationId: string;
  status: string;
}

/** finish 的响应与 POST /learning-records 同构 */
export interface LearningRecordCreated {
  recordId: string;
  subject: Subject;
  startedAt: string;
  durationMinutes: number;
  planTaskId?: string | null;
  behavior: {
    completion: Completion;
    accuracy?: number | null;
    interruptions?: number;
    blurCount?: number | null;
  };
  selfReport?: SelfReportInput | null;
  note?: string | null;
  source: 'self_report' | 'exam';
  sourceExamId?: string | null;
  assessment: {
    assessmentId?: string | null;
    subject: Subject;
    windowScore?: number | null;
    trend?: string | null;
    stateLabel: string;
    dataSufficient: boolean;
    recordCount: number;
  };
  recommendation?: RecommendationHandle | null;
  createdAt: string;
}

/** 开始计时。已有进行中会话时后端返回 409（不静默接管、也不静默丢弃）。 */
export function startTimerSession(body: TimerSessionStart): Promise<TimerSession> {
  return apiPost<TimerSession>('/timer-sessions', body);
}

/** 当前会话 + 恢复视图。进页面 / 刷新 / 换设备回来都调它。 */
export function getCurrentTimerSession(signal?: AbortSignal): Promise<TimerCurrent> {
  return apiGet<TimerCurrent>('/timer-sessions/current', undefined, signal);
}

export function getTimerSession(sessionId: string, signal?: AbortSignal): Promise<TimerSession> {
  return apiGet<TimerSession>(`/timer-sessions/${sessionId}`, undefined, signal);
}

/**
 * 上报心跳。**僵尸判定与「异常会话」识别的唯一依据**：
 * 不刷新心跳的后果不是报错，而是下次回来会被判成异常会话、要求用户裁决，所以宁可多报。
 * 前端应定时上报，并在每次用户交互时顺带上报。
 */
export function heartbeatTimerSession(sessionId: string): Promise<TimerSession> {
  return apiPost<TimerSession>(`/timer-sessions/${sessionId}/heartbeat`);
}

/** 切换任务 → 开新的一段（#14：一会话内分段，不允许并行计时）。 */
export function switchTimerTask(sessionId: string, taskId?: string | null): Promise<TimerSegment> {
  return apiPost<TimerSegment>(`/timer-sessions/${sessionId}/segments`, { taskId: taskId ?? null });
}

/**
 * 结束并收尾 —— 会话结束与落学习记录是**一个原子请求**。
 * 不在前端分两步（先结束、再 POST /learning-records）：两步之间失败会留下
 * "会话结束了但没有记录"的黑洞，正是 D31 要消灭的僵尸形态。
 */
export function finishTimerSession(
  sessionId: string,
  body: TimerFinishInput,
): Promise<LearningRecordCreated> {
  return apiPost<LearningRecordCreated>(`/timer-sessions/${sessionId}/finish`, body);
}

/** 裁决卡「丢弃」：会话标记为 abandoned，**不产生任何学习记录**。 */
export function discardTimerSession(
  sessionId: string,
): Promise<{ discarded: boolean; sessionId: string }> {
  return apiDelete<{ discarded: boolean; sessionId: string }>(`/timer-sessions/${sessionId}`);
}

// ---------- 纯函数：恢复与展示（不碰网络，便于单测与复用） ----------

/**
 * 按 mode 计算当前已过秒数。
 *
 * 服务端只给"会话起点 + 模式"，**不持续下发剩余时间**——所以前端每秒自己按
 * `startedAt` 推算，而不是维护一个递减的本地计数器。这是"刷新不丢上下文"的关键：
 * 本地计数器一刷新就归零，而按起点推算永远是对的。
 */
export function computeElapsedSeconds(
  session: Pick<TimerSession, 'startedAt'>,
  nowMs: number,
  pausedSeconds = 0,
): number {
  const startedMs = new Date(session.startedAt).getTime();
  if (Number.isNaN(startedMs)) return 0;
  const wallSeconds = (nowMs - startedMs) / 1000;
  return Math.max(0, Math.floor(wallSeconds - pausedSeconds));
}

/** 展示用：剩余（倒计时）或累计（正计时）。倒计时到点后返回 0，不显示负数。 */
export function computeDisplaySeconds(
  session: Pick<TimerSession, 'mode' | 'startedAt' | 'targetMinutes'>,
  nowMs: number,
  pausedSeconds = 0,
): { remaining: number | null; elapsed: number } {
  const elapsed = computeElapsedSeconds(session, nowMs, pausedSeconds);
  if (session.mode === 'countdown') {
    const target = (session.targetMinutes ?? 0) * 60;
    return { remaining: Math.max(0, target - elapsed), elapsed };
  }
  return { remaining: null, elapsed };
}

/** 倒计时是否已到点（到点是**软提醒**，不自动结束，D32）。 */
export function isCountdownReached(
  session: Pick<TimerSession, 'mode' | 'targetMinutes'>,
  elapsedSeconds: number,
): boolean {
  if (session.mode !== 'countdown') return false;
  return elapsedSeconds >= (session.targetMinutes ?? 0) * 60;
}
